### PR TITLE
fix(scheduler_job_runner/asset): fix how asset dag warning is added

### PR DESCRIPTION
**Pull request by Lee-W**
## Why
The logic for activating assets is not correct now.

## What
The correct logic is
1. Find the warning that should exist after this round
2. Delete the warnings that no longer needed
3. Update the warnings if they already exist and add new warnings if they do not yet exists

Related PR: https://github.com/apache/airflow/pull/43693 (that one fix the logic in the wrong way and is fixed in this PR)

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/jobs/scheduler_job_runner.py [hunk 1]
```diff
@@ -2113,44 +2114,58 @@ def _activate_referenced_assets(assets: Collection[AssetModel], *, session: Sess
...
                     continue
                 incoming_name_to_uri[asset.name] = asset.uri
                 incoming_uri_to_name[asset.uri] = asset.name
                 session.add(AssetActive.for_asset(asset))
 
-        warnings_to_have = {w.dag_id: w for w in _activate_assets_generate_warnings()}
+        def _get_first_item(x: Sequence[Any]) -> Any:
+            return x[0]
+
+        warnings_to_have = {
```
Comment (Lee-W) [inline airflow/jobs/scheduler_job_runner.py:2139]: @uranusjr, I'm not sure whether we want to warn a case like this 🤔 It kinda makes the logic a bit more complicated. The message is just concatenated.

```python
    schedule=(
        Asset(name="asset1_producer", uri="s3://bucket/asset1_producer1")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer2")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer3")
        | Asset(name="asset1_producer", uri="s3://bucket/asset1_producer4")
    ),
```

File: airflow/jobs/scheduler_job_runner.py [hunk 2]
```diff
@@ -2113,44 +2114,58 @@ def _activate_referenced_assets(assets: Collection[AssetModel], *, session: Sess
...
-                    yield from _generate_dag_warnings(asset, "uri", existing_name)
+                    yield from _generate_warning_message(asset, "uri", existing_name)
                     continue
                 incoming_name_to_uri[asset.name] = asset.uri
                 incoming_uri_to_name[asset.uri] = asset.name
                 session.add(AssetActive.for_asset(asset))
 
-        warnings_to_have = {w.dag_id: w for w in _activate_assets_generate_warnings()}
+        def _get_first_item(x: Sequence[Any]) -> Any:
+            return x[0]
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: There’s a built-in solution, `operator.itemgetter(0)`

File: airflow/jobs/scheduler_job_runner.py (hunk 1 above)
Comment (uranusjr) [reply airflow/jobs/scheduler_job_runner.py:2139]: Do you mean we should reduce the warnings from 3 to 1?

File: airflow/jobs/scheduler_job_runner.py [hunk 3]
```diff
@@ -2113,44 +2114,58 @@ def _activate_referenced_assets(assets: Collection[AssetModel], *, session: Sess
...
+
+        warnings_to_have = {
+            dag_id: DagWarning(
+                dag_id=dag_id,
+                error_type=DagWarningType.ASSET_CONFLICT,
+                message="\n".join([row[1] for row in group]),
+            )
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]:
```suggestion
            dag_id: DagWarning(
                dag_id=dag_id,
                error_type=DagWarningType.ASSET_CONFLICT,
                message="\n".join(message for _, message in group]),
            )
```

File: airflow/jobs/scheduler_job_runner.py [hunk 4]
```diff
@@ -2113,44 +2114,58 @@ def _activate_referenced_assets(assets: Collection[AssetModel], *, session: Sess
...
+            for dag_id, group in groupby(
+                sorted(_activate_assets_generate_warnings(), key=_get_first_item), key=_get_first_item
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: Actually I don’t think we even need to specify `key` here; tuple is naturally ordered by its first item anyway.

Comment (Lee-W) [reply airflow/jobs/scheduler_job_runner.py]: yep, you're right, we don't need to do it here. but I think `groupby` still need it?

File: airflow/jobs/scheduler_job_runner.py (hunk 1 above)
Comment (Lee-W) [reply airflow/jobs/scheduler_job_runner.py:2139]: In the existing logic, we only raise the latest one (if I'm not mistaken). In this implementation, the error will look like "Cannot activate asset Asset(...); ... is already associated to ...\nCannot activate asset Asset(...); ... is already associated to ...".

I don't think we can add multiple DagWarning unless we want to remove the unique constriant 🤔

File: airflow/jobs/scheduler_job_runner.py (hunk 3 above)
Comment (Lee-W) [reply airflow/jobs/scheduler_job_runner.py]: updated. Thanks!

File: airflow/jobs/scheduler_job_runner.py (hunk 4 above)
Comment (Lee-W) [reply airflow/jobs/scheduler_job_runner.py]: updated. Thanks!

File: airflow/jobs/scheduler_job_runner.py (hunk 2 above)
Comment (Lee-W) [reply airflow/jobs/scheduler_job_runner.py]: updated. Thanks!

**Review by ephraimbuddy — APPROVED** [review]

**Pull request merged**